### ISSUE TITLE
pygettext: use an AST parser instead of a tokenizer

### ISSUE THREAD

**Issue by tomasr8** [python/cpython#104400]
Follow up on this [forum discussion](https://discuss.python.org/t/modernize-and-add-missing-features-to-pygettext/26455)

This is a part 1/X of improving [pygettext](https://github.com/python/cpython/blob/main/Tools/i18n/pygettext.py). Replacing the tokenizer that powers the message extraction with a parser will simplify the code (no more counting brackets and f-string madness) and make it much easier to extend it with new features later down the road.

This change should also come with a healthy dose of new tests to verify the implementation. 

PR coming shortly ;)

### Linked PRs
* gh-104402
* gh-108173
* gh-126361
* gh-126362
* gh-129580
* gh-129672

**Comment by terryjreedy** [comment]
We replaced the custom parser in pyclbr with an ast visitor a couple of years ago.  It was shorter and clearer and agreed to be a definite improvement.  For this type of application, any possible slowdown is irrelevant.

**Comment by serhiy-storchaka** [comment]
Thanks, @tomasr8. The AST based implementation is much clearer.

The drawback is that it prevents implementing the `--add-comments` option. I think that the right solution of this is to add optional support of comments in AST -- it could also be useful in other applications. But this is a different issue.

Is it all, or you planned other work on this issue?

**Comment by AA-Turner** [comment]
Having comments in the AST would also be useful for applications like Sphinx -- is there an open issue to track this, or should we create one

A

**Comment by tomasr8** [comment]
> The drawback is that it prevents implementing the --add-comments option. I think that the right solution of this is to add optional support of comments in AST -- it could also be useful in other applications. But this is a different issue.

Yes, having the comments in the AST would be great - I can look into that and see if there are any issues open about it!

There is actually a way to add translator comments even with the AST. You can run the tokenizer separately and collect all comments and their corresponding line numbers. You can then match the comments to gettext calls. I had a working proof of concept implementation of this for [babel](https://github.com/python-babel/babel/) but I can repurpose it for pygettext!

> Is it all, or you planned other work on this issue?

Not specifically in this issue, I think we can close it. I have lots of other things I'd like to improve but I think it's better to open separate issues for that (translator comments being one of those)

**Comment by tomasr8** [comment]
> is there an open issue to track this, or should we create one

There's a similar issue about type comments: https://github.com/python/cpython/issues/101494

**Comment by sergey-miryanov** [comment]
> Yes, having the comments in the AST would be great - I can look into that and see if there are any issues open about it!

We didn't fill an issue, 'cause we stick to too old Python version, but in our applications, we need to support comments in AST and now have to do tricky re.subs to preserve them.

**Comment by serhiy-storchaka** [comment]
> There is actually a way to add translator comments even with the AST. You can run the tokenizer separately and collect all comments and their corresponding line numbers. You can then match the comments to gettext calls. I had a working proof of concept implementation of this for [babel](https://github.com/python-babel/babel/) but I can repurpose it for pygettext!

This is even more general and flexible way, but it may be hard. With comments in AST we have an issue of representing comments in a way that will not break existing code that uses AST if the tree contains comments. There is a problem of representing comments between string literals which will be concatenated into a single AST node -- it can be ignored for `pygettext`, but not for other applications. There is a problem of distinguishing comments that precede a node from a comment that follows it on the same line from comments that follow it on the following lines. Different applications may need to distinguish different types of comments and may have different preferences of the representation.

But there is no urge to add the `--add-comments` option right now.
